test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+        # assert(tx['vin'][0]['txid'] in [tx1, tx2])
+
+        # fees for 1 legacy input tx = (148+42)*20 = 3800sat
+        # at this point we're dropping to fees enough so we can spend heavier input with less waste
+        tx = select_coins_for(0.99996200)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # paying enough fees for 2x bech32 input tx
+        # but using one legacy and dropping for fees is still more efficient in terms of waste
+        tx = select_coins_for(0.99995820)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # max amount when dropping for fees is still more efficient than creating a change
+        # With fee = 4980, we'd be dropping 1180sat excess which brings waste to 2660
+        # which is equal to the waste of 2x bech32 solution
+        tx = select_coins_for(0.99995021)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 1)
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # a solution with two bech32 inputs and a change is the most efficient
+        # but it could only be found by SRD
+        tx = select_coins_for(0.99995020)
+        # TODO: randomness here, could be one or two inputs
+        # print(tx['vin'])
+        # print("{} / {}".format(len(tx['vin']), len(tx['vout'])))
+
+        # With fee at 5101sat we would be dropping 1301sat for fess
+        # This is more than cost of change and BnB won't find a solution with single input
+        # We either get
+        # 1) 2x bech32 solution (SRD only)
+        # 2) legacy + bech32 solution
+        tx = select_coins_for(0.99994899)
+        # legacy + dropping more for fees is still better in terms of waste than legacy + bech32
+        # TODO: implement optimization and update asserts
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+        # a solution with change is the best solution
+        # 2x bech32 is still more efficient than bech32-and-legacy, but only could be found with SRD
+        tx = select_coins_for(0.99994220)
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+    def test_one_big_and_many_small_coins(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_and_many_small_coins")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_many_small_coins")
+
+        # Create a bunch of small coins
+        for i in range(100):
+            w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.05)
+
+        # We will target a solution with 20 coins that will add up exactly to 1btc.
+        # Add a coin that will cover all the fees and exactly match the target.
+        # TODO: fix that knapsack targets change even if there is no change needed
```

Comment:
I'm not really arguing for improving it. I guess maybe one way that fix that would be to just remove Knapsack that targets no change and leave only knapsack that targets a MIN_CHANGE.

P.S. BnB won't produce same solutions even in the situations when a changeless solution is possible, because BnB optimizes for waste. But I think this is okay.